sotoki/sotoki.py

```diff
@@ -880,29 +900,74 @@ def get_filetype(headers, path):
     return ftype
 
 
-def download_image(url, fullpath, convert_png=False, resize=False):
-    headers = None
-    tmp_img = None
+def download_from_cache(key, output, meta_tag, meta_val, cache_storage_url):
+    cache_storage = KiwixStorage(cache_storage_url)
+    if cache_storage.has_object_matching_meta(key, meta_tag, meta_val):
+        try:
+            cache_storage.download_file(key, output, progress=True)
+            return True
+        except Exception as e:
+            print("Failed to download from cache\n" + str(e))
+            return False
+    return False
+
+
+def upload_to_cache(fpath, key, meta_tag, meta_val, cache_storage_url):
+    cache_storage = KiwixStorage(cache_storage_url)
     try:
-        tmp_img = get_tempfile(os.path.basename(fullpath))
-        headers = download(url, tmp_img, timeout=60)
-    except urllib.error.URLError as e:
-        os.unlink(tmp_img)
-        print("Cannot download " + fullpath)
-        print(e)
-    else:
-        ext = get_filetype(headers, tmp_img)
+        cache_storage.upload_file(fpath, key, meta={meta_tag: meta_val})
+    except Exception as e:
+        raise Exception("Failed to upload to cache\n" + str(e))
+
+
+def get_meta_from_url(url):
+    response_headers = requests.head(url=url).headers
+    if response_headers.get("etag") is not None:
+        return "etag", response_headers["etag"]
+    if response_headers.get("last-modified") is not None:
+        return "last-modified", response_headers["last-modified"]
+    if response_headers.get("content-length") is not None:
+        return "content-length", response_headers["content-length"]
+
+
+def download_image(url, fullpath, convert_png=False, resize=False, cache_storage=None):
```

Comment:
I don't like that `cache_storage` here refers to an URL. Could you rename it to `cache_storage_url` so that it's clear about what it is and where we pass it?

Reply:
Fixed